### PR TITLE
sign and verify by using :public_key

### PR DESCRIPTION
**Pull request by kalifs**
* Uses :public_key.sign/3 and :public_key.verify/4 to deal with signatures
* Supports Elliptic Curve Digital Signature Algorithm

### CODE REVIEW

File: lib/signex/signer.ex [hunk 1]
```diff
@@ -20,7 +25,30 @@ defmodule SignEx.Signer do
...
+    :public_key.sign(message, :sha512, decoded_pem(key))
+  end
+
+  defp decoded_pem(key) do
+    case :public_key.pem_decode(key) do
+      [_, key_entry] -> key_entry
+      [key_entry] -> key_entry
+      key_entry -> raise("unknown key type #{inspect key_entry}")
+    end
+    |> :public_key.pem_entry_decode
```
Comment (hamstah) [inline lib/signex/signer.ex]: can that fail?

File: lib/signex/signer.ex [hunk 2]
```diff
@@ -20,7 +25,30 @@ defmodule SignEx.Signer do
...
+  end
+
+  defp algorithm_from_key(key) do
+    key_type = case :public_key.pem_decode(key) do
```
Comment (hamstah) [inline lib/signex/signer.ex]: could move the case block to a function as it's used by decoded_pem too

File: test/signex_test.exs [hunk 3]
```diff
@@ -1,67 +1,141 @@
...
+        private_key: File.read!(Path.expand("../keys/private_key.pem", __ENV__.file))
+      }
+    end
 
-  describe "#sign" do
-    test "takes map as message and returns signatures and plaintext version of map" do
-      {:ok, signature, plaintext} = SignEx.sign(@message, @private_key)
+    test "takes map as message and returns signatures and plaintext version of map", %{private_key: private_key} do
+      {:ok, signature, plaintext} = SignEx.sign(@message, private_key)
       assert String.match?(signature, ~r/^Signature algorithm=rsa-sha512,key_id=bc:d8:4c:0c:eb:dd:f2:61:4e:f6:da:08:f4:b3:e0:47,salt=[\w\/\+=]+,signature=[\w\/\+=]+$/)
```
Comment (hamstah) [inline test/signex_test.exs]: Why not keep @test_signature and use it in the test?

File: test/signex_test.exs [hunk 4]
```diff
@@ -1,67 +1,141 @@
...
       assert Poison.decode!(plaintext, keys: :atoms) == @message
     end
 
-    test "takes string as message and returns signature and encoded plaintext" do
-      {:ok, signature, plaintext} = SignEx.sign(@message_str, @private_key)
+    test "takes string as message and returns signature and encoded plaintext", %{private_key: private_key} do
+      {:ok, signature, plaintext} = SignEx.sign(@message_str, private_key)
       assert String.match?(signature, ~r/^Signature algorithm=rsa-sha512,key_id=bc:d8:4c:0c:eb:dd:f2:61:4e:f6:da:08:f4:b3:e0:47,salt=[\w\/\+=]+,signature=[\w\/\+=]+$/)
```
Comment (hamstah) [inline test/signex_test.exs]: Could move the expected sig a to @ variable

File: test/signex_test.exs (hunk 3 above)
Comment (kalifs) [inline test/signex_test.exs]: they are different for different key types

**Pull request merged**